server/authentication.js

```diff
@@ -8,42 +16,196 @@ function enforceHTTPS(request, response, next) {
+    : `https://${request.headers.host}`;
+}
+
 // Only allow folks who are authorized.  Call `next` if they are authorized,
 // and return a 404 with no body if not.
-function onlyAllowResearchers(request, response, next) {
+function onlyAllowResearchers(pool, request, response, next) {
   // TODO
   // 1. Read the token from the request header
   // 2. Check to see if the token header is in `tokens` database table and it's less than 24 hours old
   // 3. This code should run in both development and production.
+
   if (process.env.NODE_ENV === 'development') return next();
-  response.status(404);
-  response.end();
+
+  const token = request.body['token'];
+  const email = request.body['email'];
+  const now = new Date();
+
+  // Check if link is in links DB
+  const sql = `
+    SELECT * 
+    FROM tokens 
+    WHERE token=$1 
+      AND email=$2
+      AND TO_TIMESTAMP($3) > timestampz
+      And TO_TIMESTAMP($3) < (timestampz + INTERVAL '24 hours')
+    ORDER BY id ASC LIMIT 1`;
+  const values = [token, email, now];
+  pool.query(sql, values).catch(err => {
+    console.log('loginEndpoint returned error');
+    console.log({ error: err });
+  });
+
+  const resultFound = true;
+  if (resultFound) {
+    return  next();
+  }
+  return sendUnauthorized(response);
+}
+
+function insertLink(pool, email, domain) {
+  const linkToken = uuid.v4();
+  //TODO: what should link look like???
+  const link = `${domain}/review_link?${qs.stringify({linkToken})}`;
+
+  // Insert link into database
+  const linkSQL = `INSERT INTO links(email, link, timestampz) VALUES ($1, $2, $3)`;
+  const now = new Date();
+  const linkValues = [email, link, now];
+  return pool.query(linkSQL, linkValues)
+    .then(results => link)
+    .catch(err => {
+      console.log('query returned err: ', err);
+      console.log({ error:err });
+    });
 }
 
+function emailLink(mailgunEnv, email, link) {
+  const linkText = link;
+  const loginlinkFilename = path.join(__dirname,'game/emails/loginlink.html.mustache');
+  const html = renderEmail(loginlinkFilename,{linkText});
+
+  const info = {
+    toEmail: email,
+    fromEmail: 'swipe-right-bot@tsl.mit.edu',
+    subject: 'Swipe Right for CS: Login Link'
+  };
+
+  if (process.env.NODE_ENV !== 'production') {
+    if (process.env.NODE_ENV === 'development') {
+      console.log('No emailing except for in production mode. Go to the following link to move forward.');
+      console.log(link);
+    }
+    return Promise.resolve();
+  }
 
-function loginEndpoint(req, res){
+  return new Promise((resolve, reject) => {
+    sendEmail(mailgunEnv, info, html, (err, mailgunResponse) => {
+      if (err) {
+        console.log("Mailgun request (err):\n  " + JSON.stringify(err, null, 2));
+        return reject(err);
+      }
+      console.log("Mailgun returned:\n  " + JSON.stringify(mailgunResponse, null, 2));
+      return resolve();   
+    });
+  });
+}
+
+
+function loginEndpoint(pool, mailgunEnv, request, response){
```

Comment:
Are these still TODOs or could these comments be updated and moved above the function to describe what it does?

Reply:
Removed the TODOs and added description of function